File: src/09.tree/26.104.ts
import { TreeNode } from "./01.100";

export function maxDepth(root: TreeNode | null): number {
    if (!root) {
        return 0;
    }
    const lMax = maxDepth(root.left);
    const rMax = maxDepth(root.right);
    return Math.max(lMax, rMax) + 1;
}

File: src/09.tree/26.104.go
package tree

func MaxDepth1(root *TreeNode) int {
	if root == nil {
		return 0
	}

	var dfs func(node *TreeNode, cur int)
	max := 1
	dfs = func(node *TreeNode, cur int) {
		if node == nil {
			return
		}

		if max < cur {
			max = cur
		}

		dfs(node.Left, cur+1)
		dfs(node.Right, cur+1)
	}

	dfs(root, 1)

	return max
}

func MaxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	lMax := MaxDepth(root.Left) + 1
	rMax := MaxDepth(root.Right) + 1

	if lMax > rMax {
		return lMax
	}

	return rMax
}
